### ISSUE TITLE
System.Net.Http.HttpClient always fails on first attempted connection

### ISSUE THREAD

**Issue by mkellen1** [dotnet/runtime#95487]
### Description

On the first attempt to connect to a web site, the connection always fails with a "Received an unexpected EOF or 0 bytes from the transport stream." exception.  On the second and subsequent attempts it works as expected. 

This was first apparent to me because the behavior appeared suddenly in PowerShell 7.2.x+.  A simple Invoke-WebRequest fails the first time, then works fine for the rest of the session. Meanwhile, PowerShell 5 works as expected on the same platform.

Discussion over in the PowerShell repo in topic 20813 led to trying a purely C# code test (below) that suggests this is not a PowerShell issue, but some weird edge case that is surfaced in .NET  

I can reproduce the behavior on multiple systems in my organization, on the LAN or not, even completely disconnected from the network using a loopback web server. I can also reproduce it under Windows Subsystem for Linux (kali) using Mono (6.8.0.105). 

### Reproduction Steps

```c#
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Net.Http;
using System.Net;


namespace HTTP_Test
{
    class program
    {
        static void Main()
        {
            Task t = new Task(HTTP_GET);
            t.Start();
            Console.ReadLine();
        }

        static async void HTTP_GET()
        {
            var TARGETURL = "https://www.google.com/";

            HttpClientHandler handler = new HttpClientHandler();

            Console.WriteLine("GET: + " + TARGETURL);

            // ... Use HttpClient.
            HttpClient client = new HttpClient(handler);

            HttpResponseMessage response = await client.GetAsync(TARGETURL);
            HttpContent content = response.Content;

            // ... Check Status Code
            Console.WriteLine("Response StatusCode: " + (int)response.StatusCode);

        }
    }
}
```

### Expected behavior

Network connection should work the first time.

### Actual behavior

Connection fails on first attempt every time.

### Known Workarounds

Use PowerShell 5.

### Configuration

My system is .NET release 533325, Windows 10 22H2 (19045.3570) on an x64.

Has been reproduced on multiple versions of Windows 10 on various hardware platforms, including a VDI.  This has also been reproduced under Windows Subsystem for Linux on my laptop.

**Comment by CarnaViire** [comment]
@mkellen1 we weren't able to reproduce it so far. Could you please provide a full runnable repo (together with the loopback server you mentioned)?

Packet captures would be helpful as well.

**Comment by CarnaViire** [comment]
Also, can you please confirm which .NET version are you using for the repro application? Is it .NET 8.0?

Can you check the same for .NET 6 and .NET 5? I have a suspicion that's related to HttpClient's retry policy that has changed since .NET 6 (https://github.com/dotnet/runtime/issues/60644)

**Comment by mklement0** [comment]
Just to link this to the original discussion in the PowerShell repo: 

* https://github.com/dotnet/runtime/issues/95487

**Comment by wfurt** [comment]
can you get packet captures @mklement0? This looks environmental and I think it is going to be difficult without it.

**Comment by mklement0** [comment]
@wfurt, I'm just a liaison here, I think you meant to address the OP, @mkellen1.

**Comment by mkellen1** [comment]
There are no packets captured by wireshark for the first connection attempt.  The second connection attempt looks normal. (Ignore the RST at the end, that was me killing the program).
[capture.zip](https://github.com/dotnet/runtime/files/13530462/capture.zip)

I will see if I can set up the additional SDKs to try with older versions. I have confirmed that it works the same directly using the .Net 8 SDK (8.0.100)

**Comment by mkellen1** [comment]
There were no packets generated to capture.  I don't know what you expect me to do about that.

**Comment by mkellen1** [comment]
The systems have Palo Alto's Cortex XDR for endpoint protection (no alerts noted in the console) and Forcepoint ONE client for web filtering (no blocking noted in the console). I see the same behavior on prem or remote, with or without VPN.

**Comment by wfurt** [comment]
it may be bug in connection pool logic @CarnaViire mentioned. Perhaps she can comment on...? 
Disabling protection (for the test) or running on different system still may be worth off try.

**Comment by mkellen1** [comment]
Disabling each in turn made no difference (nor did turning them both off together).

**Comment by ManickaP** [comment]
Hey @mkellen1 could please also try what @CarnaViire suggested:
> Can you check the same for .NET 6 and .NET 5? I have a suspicion that's related to HttpClient's retry policy that has changed since .NET 6 (https://github.com/dotnet/runtime/issues/60644)

So that we know if it might be related to that change.

**Comment by CarnaViire** [comment]
If we confirm that the behavior changes between .NET 5 and .NET 6, it will shed more light on why the error only started to show in newer pwsh versions - It's because it was hidden before. (However, it will still not fully explain the scenario and why the server sends the RST. BTW I believe we _should_ see this first RST on the capture...(

It is not a bug @wfurt, we previously were not RFC-compliant and retried too much. Now it's fixed. Currently I don't believe there's any problem on .NET side. But it still would be good to understand the scenario fully.

**Comment by mkellen1** [comment]
Well, darnit, I had hoped that it was the .NET 6 change, but I can reproduce the issue with both 6.0.417 and 5.0.408

**Comment by mkellen1** [comment]
I went back and downloaded Powershell versions until I found exactly where it broke. The last version that works for me is 7.1.0-preview.7.  The next version (7.1.0-rc.1) fails. Each had a different release of .NET 5.

Sure enough, the code works in dotnet-5.0.100-preview.8 and fails in dotnet-5.0.100-rc1, matching the powershell breakage.

**Comment by CarnaViire** [comment]
*TL;DR: Here's a modified test case based on yours, please run it (verify that the issue happens) and share the output with us.*
[HttpTest-5.0-p8.zip](https://github.com/dotnet/runtime/files/13666589/HttpTest-5.0-p8.zip)

Hi @mkellen1, sorry for the delay. Can I please ask you to also collect the extended telemetry while the issue happens? It is very strange that there are no packages for the first connection attempt, but we should at least see it in the telemetry.

Official docs for networking telemetry are [here](https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/telemetry/events), but we also need internal diagnostic event sources which are not listed on this page; also, as a funny coincidence, the names of these internal event sources changed right between 5.0-p8 and 5.0-rc1 😄 (https://github.com/dotnet/runtime/commit/ba7c518a494ef7aa87159816e07576eefe14290f) -- so I went ahead and wrote for you an event listener class to collect everything:

<details>
<summary>NetEventListener</summary>

```c#
    public sealed class NetEventListener : EventListener
    {
        public static string[] NetworkingEvents => new[]
        {
            "System.Net.Http",
            "System.Net.NameResolution",
            "System.Net.Sockets",
            "System.Net.Security",
            // internal telemetry names from 5.0 rc1
            "Private.InternalDiagnostics.System.Net.Http",
            "Private.InternalDiagnostics.System.Net.NameResolution",
            "Private.InternalDiagnostics.System.Net.Sockets",
            "Private.InternalDiagnostics.System.Net.Security",
            "Private.InternalDiagnostics.System.Net.NetworkInformation",
            "Private.InternalDiagnostics.System.Net.Primitives",
            // internal telemetry names prior to 5.0 rc1
            "Microsoft-System-Net-Http",
            "Microsoft-System-Net-NameResolution",
            "Microsoft-System-Net-Sockets",
            "Microsoft-System-Net-Security",
            "Microsoft-System-Net-NetworkInformation",
            "Microsoft-System-Net-Primitives",
        };

        private readonly HashSet<string> _sourceNames;
        private List<EventSource> _eventSources = new List<EventSource>();
        public NetEventListener()
        {
            lock (this)
            {
                _sourceNames = new HashSet<string>(NetworkingEvents);
                foreach (EventSource eventSource in _eventSources)
                {
                    OnEventSourceCreated(eventSource);
                }
                _eventSources = null!;
            }
        }

        protected override void OnEventSourceCreated(EventSource eventSource)
        {
            lock (this)
            {
                // We're called from base ctor, just save the event source for later initialization.
                if (_sourceNames is null)
                {
                    _eventSources.Add(eventSource);
                    return;
                }

                // Second pass called from our ctor, allow logging for specified source names.
                if (_sourceNames.Contains(eventSource.Name))
                {
                    EnableEvents(eventSource, EventLevel.LogAlways);
                }
            }
        }

        protected override void OnEventWritten(EventWrittenEventArgs eventData)
        {
            StringBuilder sb = new StringBuilder().
                Append($"{eventData.TimeStamp:HH:mm:ss.fffffff}[{eventData.EventName}] ");

            for (int i = 0; i < eventData.Payload?.Count; i++)
            {
                if (i > 0)
                    sb.Append(", ");
                sb.Append(eventData.PayloadNames?[i]).Append(": ").Append(eventData.Payload[i]);
            }
            try
            {
                Console.WriteLine(sb.ToString());
            }
            catch { }
        }
    }
```

</details>

Then you need to create an instance at the start of your test app. Here's the modified test loosely based on what you've sent above:

<details>
<summary>Program</summary>

```c#
    class Program
    {
        static async Task Main(string[] args)
        {
            var listener = new NetEventListener();

            Console.WriteLine(RuntimeInformation.FrameworkDescription);
            var url = "http://localhost/";

            var client = new HttpClient();
            
            Console.WriteLine("\n#1");
            await TestGetAsync(client, url);

            Console.WriteLine("\n#2");
            await TestGetAsync(client, url);
        }

        private static async Task TestGetAsync(HttpClient client, string url)
        {
            Console.WriteLine($"GET {url}\n");
            var response = await client.GetAsync(url);
            Console.WriteLine($"\n{(int)response.StatusCode} {response.StatusCode}");
            Console.WriteLine(await response.Content.ReadAsStringAsync());
        }
    }
```

</details>

As a reference, this is my output from it:

<details>
<summary>5.0 Preview 8</summary>

```
➜ HttpTest-5.0-p8  dotnet run -c release
.NET 5.0.0-preview.8.20407.11
22:42:52.8397503[Info] thisOrContextObject: HttpClientHandler#32854180, memberName: .ctor, message: 
22:42:52.8480122[Info] thisOrContextObject: SocketsHttpHandler#27252167, memberName: .ctor, message: 
22:42:52.8481729[Info] thisOrContextObject: DiagnosticsHandler#43942917, memberName: .ctor, message: 
22:42:52.8483628[Associate] thisOrContextObject: DiagnosticsHandler#43942917, memberName: InnerHandler, first: DiagnosticsHandler#43942917, second: SocketsHttpHandler#27252167
22:42:52.8501734[Associate] thisOrContextObject: HttpClient#59941933, memberName: .ctor, first: HttpClient#59941933, second: HttpClientHandler#32854180

#1
GET http://localhost/

22:42:52.8892802[Info] thisOrContextObject: WinInetProxyHelper#2606490, memberName: .ctor, message: AutoConfigUrl=(null), AutoDetect=True, Proxy=(null), ProxyBypass=(null)
22:42:52.8894258[Info] thisOrContextObject: WinInetProxyHelper#2606490, memberName: .ctor, message: _useProxy=True
22:42:52.8895426[Info] thisOrContextObject: WinInetProxyHelper#2606490, memberName: TryCreate, message: AutoSettingsUsed, calling WinHttpOpen
22:42:52.8960012[Info] thisOrContextObject: HttpConnectionHandler#23458411, memberName: .ctor, message: 
22:42:52.8961003[Info] thisOrContextObject: RedirectHandler#9799115, memberName: .ctor, message: 
22:42:52.8977850[RequestStart] scheme: http, host: localhost, port: 80, pathAndQuery: /, versionMajor: 1, versionMinor: 1
22:42:52.9006364[ErrorMessage] thisOrContextObject: WinInetProxyHelper#2606490, memberName: GetProxyForUrl, message: error=12180
22:42:52.9008192[Info] thisOrContextObject: WinInetProxyHelper#2606490, memberName: GetProxyForUrl, message: useProxy=False
22:42:52.9029805[HandlerMessage] poolId: 21083178, workerId: 0, requestId: 0, memberName: .ctor, message: HttpConnectionPool http://localhost:80
22:42:52.9059360[HandlerMessage] poolId: 21083178, workerId: 0, requestId: 0, memberName: IncrementConnectionCountNoLock, message: 
22:42:52.9061533[HandlerMessage] poolId: 21083178, workerId: 0, requestId: 0, memberName: GetHttpConnectionAsync, message: Creating new connection for pool.
22:42:52.9108481[Info] thisOrContextObject: ConnectEventArgs#55530882, memberName: InitializeInternals, message: new PreAllocatedOverlapped PreAllocatedOverlapped#30015890
22:42:52.9179918[Info] thisOrContextObject: (null), memberName: GetHostName, message: knatalia-P53
22:42:52.9207439[Info] thisOrContextObject: Socket#1707556, memberName: .ctor, message: InterNetwork
22:42:52.9213164[Info] thisOrContextObject: (null), memberName: CreateSocket, message: SafeSocketHandle:15368010(0x524)
22:42:52.9214536[Info] thisOrContextObject: Socket#4094363, memberName: .ctor, message: InterNetworkV6
22:42:52.9216312[Info] thisOrContextObject: (null), memberName: CreateSocket, message: SafeSocketHandle:36849274(0x528)
22:42:52.9317561[Info] thisOrContextObject: SocketAsyncEventArgs#6044116, memberName: InitializeInternals, message: new PreAllocatedOverlapped PreAllocatedOverlapped#59817589
22:42:52.9323020[Info] thisOrContextObject: Socket#4094363, memberName: WildcardBindForConnectIfNecessary, message: ::
22:42:52.9328131[ConnectStart] address: InterNetworkV6:28:{0,80,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0}
22:42:52.9333013[Info] thisOrContextObject: SafeSocketHandle#36849274, memberName: GetOrAllocateThreadPoolBoundHandle, message: calling ThreadPool.BindHandle()
22:42:52.9390887[Connected] localEp: IPEndPoint#-508592849, remoteEp: (null), socketHash: 4094363
22:42:52.9393856[ConnectStop]
22:42:52.9396242[Info] thisOrContextObject: Socket#4094363, memberName: SetToConnected, message: now connected
22:42:52.9399142[Info] thisOrContextObject: Socket#1707556, memberName: Connected, message: _isConnected:False
22:42:52.9400941[Info] thisOrContextObject: Socket#1707556, memberName: Dispose, message: timeout = -1
22:42:52.9403678[Info] thisOrContextObject: Socket#1707556, memberName: Dispose, message: disposing:True Disposed:False
22:42:52.9406386[Info] thisOrContextObject: Socket#1707556, memberName: Dispose, message: Calling _handle.CloseAsIs()
22:42:52.9408244[Info] thisOrContextObject: SafeSocketHandle#15368010, memberName: CloseAsIs, message: shouldClose=True
22:42:52.9410259[Info] thisOrContextObject: SafeSocketHandle#15368010, memberName: ReleaseHandle, message: shouldClose=False
22:42:52.9411965[Info] thisOrContextObject: SafeSocketHandle#15368010, memberName: CloseHandle, message: handle:0x524
22:42:52.9414273[Info] thisOrContextObject: SafeSocketHandle#15368010, memberName: DoCloseHandle, message: handle:0x524, Following 'blockable' branch
22:42:52.9416980[Info] thisOrContextObject: SafeSocketHandle#15368010, memberName: DoCloseHandle, message: handle:0x524, closesocket()#1:Success
22:42:52.9422218[Info] thisOrContextObject: Socket#4094363, memberName: Connected, message: _isConnected:True
22:42:52.9430180[Info] thisOrContextObject: Socket#4094363, memberName: SetSocketOption, message: optionLevel:Tcp optionName:NoDelay optionValue:1
22:42:52.9436246[Info] thisOrContextObject: Socket#4094363, memberName: SetSocketOption, message: SetSockOpt returns errorCode:Success
22:42:52.9438701[Info] thisOrContextObject: Socket#4094363, memberName: Connected, message: _isConnected:True
22:42:52.9456343[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 0, memberName: TraceConnection, message: HttpConnection(HttpConnectionPool http://localhost:80)
22:42:52.9486705[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 11429296, memberName: SendAsyncCore, message: Sending request: Method: GET, RequestUri: 'http://localhost/', Version: 1.1, Content: <null>, Headers:
{
}
22:42:52.9506580[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 11429296, memberName: WriteToStreamAsync, message: Writing 35 bytes.
22:42:52.9509645[Info] thisOrContextObject: AwaitableSocketAsyncEventArgs#41622463, memberName: InitializeInternals, message: new PreAllocatedOverlapped PreAllocatedOverlapped#31364015
22:42:52.9521955[DumpBuffer] thisOrContextObject: AwaitableSocketAsyncEventArgs#41622463, memberName: LogBuffer, buffer: System.Byte[]
22:42:52.9525945[Associate] thisOrContextObject: HttpResponseMessage#28379535, memberName: RequestMessage, first: HttpResponseMessage#28379535, second: HttpRequestMessage#11429296
22:42:52.9528216[Info] thisOrContextObject: HttpConnectionResponseContent#64109423, memberName: .ctor, message:
22:42:52.9530812[Associate] thisOrContextObject: HttpResponseMessage#28379535, memberName: Content, first: HttpResponseMessage#28379535, second: HttpConnectionResponseContent#64109423
22:42:52.9541727[Info] thisOrContextObject: AwaitableSocketAsyncEventArgs#35236192, memberName: InitializeInternals, message: new PreAllocatedOverlapped PreAllocatedOverlapped#21943666
22:42:52.9547879[DumpBuffer] thisOrContextObject: AwaitableSocketAsyncEventArgs#35236192, memberName: LogBuffer, buffer: System.Byte[]
22:42:52.9550743[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 11429296, memberName: FillAsync, message: Received 272 bytes.
22:42:52.9613313[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 11429296, memberName: SendAsyncCore, message: Request is fully sent.
22:42:52.9647602[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 11429296, memberName: SendAsyncCore, message: Received response: StatusCode: 200, ReasonPhrase: 'OK', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
{
  Date: Wed, 13 Dec 2023 22:42:52 GMT
  Server: Apache/2.4.58
  Server: (Win64)
  ETag: "2e-60c69cce8b49a"
  Accept-Ranges: bytes
  Last-Modified: Wed, 13 Dec 2023 20:13:29 GMT
  Content-Length: 46
  Content-Type: text/html
}
22:42:52.9682997[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 11429296, memberName: CopyFromBufferAsync, message: Copying 46 bytes to stream.
22:42:52.9686556[RequestStop]
22:42:52.9693077[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 0, memberName: ReturnConnection, message: Stored connection in pool.
22:42:52.9698588[ClientSendCompleted] responseString: StatusCode: 200, ReasonPhrase: 'OK', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
{
  Date: Wed, 13 Dec 2023 22:42:52 GMT
  Server: Apache/2.4.58
  Server: (Win64)
  ETag: "2e-60c69cce8b49a"
  Accept-Ranges: bytes
  Last-Modified: Wed, 13 Dec 2023 20:13:29 GMT
  Content-Length: 46
  Content-Type: text/html
}, httpRequestMessageHash: 11429296, httpResponseMessageHash: 28379535, httpClientHash: 59941933

200 OK
<html><body><h1>It works!</h1></body></html>


#2
GET http://localhost/

22:42:52.9743191[RequestStart] scheme: http, host: localhost, port: 80, pathAndQuery: /, versionMajor: 1, versionMinor: 1
22:42:52.9752029[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 0, memberName: GetOrReserveHttp11ConnectionAsync, message: Found usable connection in pool.
22:42:52.9754916[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 26517107, memberName: SendAsyncCore, message: Sending request: Method: GET, RequestUri: 'http://localhost/', Version: 1.1, Content: <null>, Headers:
{
}
22:42:52.9758922[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 26517107, memberName: WriteToStreamAsync, message: Writing 35 bytes.
22:42:52.9761684[DumpBuffer] thisOrContextObject: AwaitableSocketAsyncEventArgs#41622463, memberName: LogBuffer, buffer: System.Byte[]
22:42:52.9770676[DumpBuffer] thisOrContextObject: AwaitableSocketAsyncEventArgs#35236192, memberName: LogBuffer, buffer: System.Byte[]
22:42:52.9782194[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 26517107, memberName: SendAsyncCore, message: Received 272 bytes.
22:42:52.9784980[Associate] thisOrContextObject: HttpResponseMessage#48209832, memberName: RequestMessage, first: HttpResponseMessage#48209832, second: HttpRequestMessage#26517107
22:42:52.9790274[Info] thisOrContextObject: HttpConnectionResponseContent#5773521, memberName: .ctor, message:
22:42:52.9792394[Associate] thisOrContextObject: HttpResponseMessage#48209832, memberName: Content, first: HttpResponseMessage#48209832, second: HttpConnectionResponseContent#5773521
22:42:52.9795244[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 26517107, memberName: SendAsyncCore, message: Request is fully sent.
22:42:52.9798857[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 26517107, memberName: SendAsyncCore, message: Received response: StatusCode: 200, ReasonPhrase: 'OK', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
{
  Date: Wed, 13 Dec 2023 22:42:52 GMT
  Server: Apache/2.4.58
  Server: (Win64)
  ETag: "2e-60c69cce8b49a"
  Accept-Ranges: bytes
  Last-Modified: Wed, 13 Dec 2023 20:13:29 GMT
  Content-Length: 46
  Content-Type: text/html
}
22:42:52.9812620[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 26517107, memberName: CopyFromBufferAsync, message: Copying 46 bytes to stream.
22:42:52.9815394[RequestStop]
22:42:52.9818202[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 0, memberName: ReturnConnection, message: Stored connection in pool.
22:42:52.9820452[ClientSendCompleted] responseString: StatusCode: 200, ReasonPhrase: 'OK', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
{
  Date: Wed, 13 Dec 2023 22:42:52 GMT
  Server: Apache/2.4.58
  Server: (Win64)
  ETag: "2e-60c69cce8b49a"
  Accept-Ranges: bytes
  Last-Modified: Wed, 13 Dec 2023 20:13:29 GMT
  Content-Length: 46
  Content-Type: text/html
}, httpRequestMessageHash: 26517107, httpResponseMessageHash: 48209832, httpClientHash: 59941933

200 OK
<html><body><h1>It works!</h1></body></html>
```

</details>

<details>
<summary>5.0 RC 1</summary>

```
➜ HttpTest-5.0-rc1  dotnet run -c release
.NET 5.0.0-rc.1.20451.14
22:44:42.9851982[Info] thisOrContextObject: HttpClientHandler#32854180, memberName: .ctor, message: 
22:44:43.0062352[Info] thisOrContextObject: SocketsHttpHandler#27252167, memberName: .ctor, message:
22:44:43.0076672[Info] thisOrContextObject: DiagnosticsHandler#43942917, memberName: .ctor, message:
22:44:43.0079784[Associate] thisOrContextObject: DiagnosticsHandler#43942917, memberName: InnerHandler, first: DiagnosticsHandler#43942917, second: SocketsHttpHandler#27252167
22:44:43.0104055[Associate] thisOrContextObject: HttpClient#59941933, memberName: .ctor, first: HttpClient#59941933, second: HttpClientHandler#32854180

#1
GET http://localhost/

22:44:43.0182075[RequestStart] scheme: http, host: localhost, port: 80, pathAndQuery: /, versionMajor: 1, versionMinor: 1, versionPolicy: 0
22:44:43.0555965[Info] thisOrContextObject: WinInetProxyHelper#2606490, memberName: .ctor, message: AutoConfigUrl=(null), AutoDetect=True, Proxy=(null), ProxyBypass=(null)
22:44:43.0558503[Info] thisOrContextObject: WinInetProxyHelper#2606490, memberName: .ctor, message: _useProxy=True
22:44:43.0560485[Info] thisOrContextObject: WinInetProxyHelper#2606490, memberName: TryCreate, message: AutoSettingsUsed, calling WinHttpOpen
22:44:43.0614976[Info] thisOrContextObject: HttpConnectionHandler#23458411, memberName: .ctor, message:
22:44:43.0617260[Info] thisOrContextObject: RedirectHandler#9799115, memberName: .ctor, message:
22:44:43.0656138[ErrorMessage] thisOrContextObject: WinInetProxyHelper#2606490, memberName: GetProxyForUrl, message: error=12180
22:44:43.0658231[Info] thisOrContextObject: WinInetProxyHelper#2606490, memberName: GetProxyForUrl, message: useProxy=False
22:44:43.0676869[HandlerMessage] poolId: 21083178, workerId: 0, requestId: 0, memberName: .ctor, message: HttpConnectionPool http://localhost:80
22:44:43.0708068[HandlerMessage] poolId: 21083178, workerId: 0, requestId: 0, memberName: IncrementConnectionCountNoLock, message:
22:44:43.0710723[HandlerMessage] poolId: 21083178, workerId: 0, requestId: 0, memberName: GetHttpConnectionAsync, message: Creating new connection for pool.
22:44:43.0765486[Info] thisOrContextObject: Socket#55530882, memberName: .ctor, message: InterNetworkV6
22:44:43.0796413[ResolutionStart] hostNameOrAddress: 
22:44:43.0863246[ResolutionStop]
22:44:43.0870359[Info] thisOrContextObject: (null), memberName: GetHostName, message: knatalia-P53
22:44:43.0879841[Info] thisOrContextObject: (null), memberName: CreateSocket, message: SafeSocketHandle:30015890(0x3FC)
22:44:43.0883055[Info] thisOrContextObject: Socket#55530882, memberName: SetSocketOption, message: optionLevel:Tcp optionName:NoDelay optionValue:1
22:44:43.0889465[Info] thisOrContextObject: Socket#55530882, memberName: SetSocketOption, message: SetSockOpt returns errorCode:Success
22:44:43.0891704[Info] thisOrContextObject: Socket#55530882, memberName: SetSocketOption, message: optionLevel:IPv6 optionName:IPv6Only optionValue:0
22:44:43.0894009[Info] thisOrContextObject: Socket#55530882, memberName: SetSocketOption, message: SetSockOpt returns errorCode:Success
22:44:43.0910378[Info] thisOrContextObject: AwaitableSocketAsyncEventArgs#1707556, memberName: InitializeInternals, message: new PreAllocatedOverlapped PreAllocatedOverlapped#15368010
22:44:43.0913709[ConnectedAsyncDns] socketHash: 55530882
22:44:43.0933471[ResolutionStart] hostNameOrAddress: localhost
22:44:43.1041238[ResolutionStop] 
22:44:43.1045104[Info] thisOrContextObject: SocketAsyncEventArgs#6044116, memberName: InitializeInternals, message: new PreAllocatedOverlapped PreAllocatedOverlapped#59817589
22:44:43.1050499[Info] thisOrContextObject: Socket#55530882, memberName: WildcardBindForConnectIfNecessary, message: ::
22:44:43.1060531[ConnectStart] address: InterNetworkV6:28:{0,80,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0}
22:44:43.1063909[Info] thisOrContextObject: SafeSocketHandle#30015890, memberName: GetOrAllocateThreadPoolBoundHandle, message: calling ThreadPool.BindHandle()
22:44:43.1108188[Connected] localEp: IPEndPoint#1150482023, remoteEp: (null), socketHash: 55530882
22:44:43.1110389[ConnectStop]
22:44:43.1112749[Info] thisOrContextObject: Socket#55530882, memberName: SetToConnected, message: now connected
22:44:43.1138452[Info] thisOrContextObject: Socket#55530882, memberName: Connected, message: _isConnected:True
22:44:43.1140932[ConnectionEstablished] versionMajor: 1, versionMinor: 1
22:44:43.1143803[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 0, memberName: TraceConnection, message: HttpConnection(HttpConnectionPool http://localhost:80)
22:44:43.1170835[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 11429296, memberName: SendAsyncCore, message: Sending request: Method: GET, RequestUri: 'http://localhost/', Version: 1.1, Content: <null>, Headers:
{
}
22:44:43.1176409[RequestHeadersStart]
22:44:43.1189413[RequestHeadersStop]
22:44:43.1192126[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 11429296, memberName: WriteToStreamAsync, message: Writing 35 bytes.
22:44:43.1194626[Info] thisOrContextObject: AwaitableSocketAsyncEventArgs#41622463, memberName: InitializeInternals, message: new PreAllocatedOverlapped PreAllocatedOverlapped#31364015
22:44:43.1203271[DumpBuffer] thisOrContextObject: AwaitableSocketAsyncEventArgs#41622463, memberName: LogBuffer, buffer: System.Byte[]
22:44:43.1205803[Associate] thisOrContextObject: HttpResponseMessage#28379535, memberName: RequestMessage, first: HttpResponseMessage#28379535, second: HttpRequestMessage#11429296
22:44:43.1207375[Info] thisOrContextObject: HttpConnectionResponseContent#64109423, memberName: .ctor, message:
22:44:43.1208491[Associate] thisOrContextObject: HttpResponseMessage#28379535, memberName: Content, first: HttpResponseMessage#28379535, second: HttpConnectionResponseContent#64109423
22:44:43.1352238[DumpBuffer] thisOrContextObject: AwaitableSocketAsyncEventArgs#1707556, memberName: LogBuffer, buffer: System.Byte[]
22:44:43.1358748[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 11429296, memberName: FillAsync, message: Received 272 bytes.
22:44:43.1371597[ResponseHeadersStart]
22:44:43.1410291[ResponseHeadersStop]
22:44:43.1413229[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 11429296, memberName: SendAsyncCore, message: Request is fully sent.
22:44:43.1445998[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 11429296, memberName: SendAsyncCore, message: Received response: StatusCode: 200, ReasonPhrase: 'OK', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
{
  Date: Wed, 13 Dec 2023 22:44:43 GMT
  Server: Apache/2.4.58
  Server: (Win64)
  ETag: "2e-60c69cce8b49a"
  Accept-Ranges: bytes
  Last-Modified: Wed, 13 Dec 2023 20:13:29 GMT
  Content-Length: 46
  Content-Type: text/html
}
22:44:43.1467810[ResponseContentStart]
22:44:43.1476227[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 11429296, memberName: CopyFromBufferAsync, message: Copying 46 bytes to stream.
22:44:43.1481601[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 0, memberName: ReturnConnection, message: Stored connection in pool.
22:44:43.1486104[ClientSendCompleted] responseString: StatusCode: 200, ReasonPhrase: 'OK', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
{
  Date: Wed, 13 Dec 2023 22:44:43 GMT
  Server: Apache/2.4.58
  Server: (Win64)
  ETag: "2e-60c69cce8b49a"
  Accept-Ranges: bytes
  Last-Modified: Wed, 13 Dec 2023 20:13:29 GMT
  Content-Length: 46
  Content-Type: text/html
}, httpRequestMessageHash: 11429296, httpResponseMessageHash: 28379535, httpClientHash: 59941933
22:44:43.1495825[ResponseContentStop]
22:44:43.1497322[RequestStop]

200 OK
<html><body><h1>It works!</h1></body></html>


#2
GET http://localhost/

22:44:43.1518377[RequestStart] scheme: http, host: localhost, port: 80, pathAndQuery: /, versionMajor: 1, versionMinor: 1, versionPolicy: 0
22:44:43.1523636[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 0, memberName: GetOrReserveHttp11ConnectionAsync, message: Found usable connection in pool.
22:44:43.1526806[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 58998806, memberName: SendAsyncCore, message: Sending request: Method: GET, RequestUri: 'http://localhost/', Version: 1.1, Content: <null>, Headers:
{
}
22:44:43.1530349[RequestHeadersStart]
22:44:43.1531862[RequestHeadersStop]
22:44:43.1533902[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 58998806, memberName: WriteToStreamAsync, message: Writing 35 bytes.
22:44:43.1536302[DumpBuffer] thisOrContextObject: AwaitableSocketAsyncEventArgs#41622463, memberName: LogBuffer, buffer: System.Byte[]
22:44:43.1547793[DumpBuffer] thisOrContextObject: AwaitableSocketAsyncEventArgs#1707556, memberName: LogBuffer, buffer: System.Byte[]
22:44:43.1550428[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 58998806, memberName: SendAsyncCore, message: Received 272 bytes.
22:44:43.1552407[Associate] thisOrContextObject: HttpResponseMessage#803548, memberName: RequestMessage, first: HttpResponseMessage#803548, second: HttpRequestMessage#58998806
22:44:43.1554858[Info] thisOrContextObject: HttpConnectionResponseContent#26517107, memberName: .ctor, message:
22:44:43.1556766[Associate] thisOrContextObject: HttpResponseMessage#803548, memberName: Content, first: HttpResponseMessage#803548, second: HttpConnectionResponseContent#26517107
22:44:43.1558318[ResponseHeadersStart]
22:44:43.1559383[ResponseHeadersStop]
22:44:43.1561241[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 58998806, memberName: SendAsyncCore, message: Request is fully sent.
22:44:43.1563498[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 58998806, memberName: SendAsyncCore, message: Received response: StatusCode: 200, ReasonPhrase: 'OK', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
{
  Date: Wed, 13 Dec 2023 22:44:43 GMT
  Server: Apache/2.4.58
  Server: (Win64)
  ETag: "2e-60c69cce8b49a"
  Accept-Ranges: bytes
  Last-Modified: Wed, 13 Dec 2023 20:13:29 GMT
  Content-Length: 46
  Content-Type: text/html
}
22:44:43.1575909[ResponseContentStart] 
22:44:43.1577866[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 58998806, memberName: CopyFromBufferAsync, message: Copying 46 bytes to stream.
22:44:43.1580002[HandlerMessage] poolId: 21083178, workerId: 12547953, requestId: 0, memberName: ReturnConnection, message: Stored connection in pool.
22:44:43.1582698[ClientSendCompleted] responseString: StatusCode: 200, ReasonPhrase: 'OK', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
{
  Date: Wed, 13 Dec 2023 22:44:43 GMT
  Server: Apache/2.4.58
  Server: (Win64)
  ETag: "2e-60c69cce8b49a"
  Accept-Ranges: bytes
  Last-Modified: Wed, 13 Dec 2023 20:13:29 GMT
  Content-Length: 46
  Content-Type: text/html
}, httpRequestMessageHash: 58998806, httpResponseMessageHash: 803548, httpClientHash: 59941933
22:44:43.1595211[ResponseContentStop]
22:44:43.1596802[RequestStop]

200 OK
<html><body><h1>It works!</h1></body></html>
```

</details>

The server is also a default Apache listening on localhost.

The whole project for the preview 8 version is in [HttpTest-5.0-p8.zip](https://github.com/dotnet/runtime/files/13666589/HttpTest-5.0-p8.zip). For rc1, you'd need to change SDK in global.json:

```diff
{
  "sdk": {
-   "version": "5.0.100-preview.8.20417.9",
+   "version": "5.0.100-rc.1.20452.10",
    "rollForward": "disable"
  }
}
```

I don't know whether you already leveraging local .NET installations, but just in case, here's how to add it (example is for Windows, but there are also docs for Linux in https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-install-script):

```
cd <path-to-projects>\HttpTest-5.0-p8
<path-to-script>\dotnet-install.ps1 -Version 5.0.100-preview.8.20417.9 -i '.\.dotnet\'
```

To verify, you can call
```
➜  HttpTest-5.0-p8 dotnet --info
.NET SDK (reflecting any global.json):
 Version:   5.0.100-preview.8.20417.9 <--------
 Commit:    fc62663a35

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\.....\HttpTest-5.0-p8\.dotnet\sdk\5.0.100-preview.8.20417.9\ <--------

Host (useful for support):
  Version: 5.0.0-preview.8.20407.11 <--------
  Commit:  bf456654f9

.....rest omitted for brevity
```

For RC 1, it would be `-Version 5.0.100-rc.1.20452.10` instead

**Comment by mkellen1** [comment]
Many thanks for the detailed response.  It may take me a bit to implement it all, but I will get back to you as soon as I can with the results.

**Comment by mkellen1** [comment]
The new test program still reproduces the error, so I went ahead and ran it once using p8 (showing it work there) and twice under rc1 (showing the initial failure and subsequent success).  Output is attached.

[p8.txt](https://github.com/dotnet/runtime/files/13689971/p8.txt)
[rc1-run1.txt](https://github.com/dotnet/runtime/files/13689974/rc1-run1.txt)
[rc1-run2.txt](https://github.com/dotnet/runtime/files/13689975/rc1-run2.txt)

**Comment by MihaZupan** [comment]
One notable difference I see in the traces is that we're connecting to an IPv4 endpoint on preview8, but IPv6 in rc1.

A lot of networking changes happened [during that time](https://github.com/dotnet/runtime/pulls?q=is%3Apr+is%3Aclosed+merged%3A2020-07-18..2020-09-01+label%3Aarea-System.Net%2Carea-System.Net.Sockets%2Carea-System.Net.Http), including ones that meant we started using dual-mode sockets when connecting by default AFAICT (https://github.com/dotnet/runtime/issues/44686#issuecomment-733845701, [Preview 8 code](https://github.com/dotnet/runtime/blob/bf456654f9a4f9a86c15d9d50095ff29cde5f0a4/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs#L55), [RC1 code](https://github.com/dotnet/runtime/blob/38017c3935de95d0335bac04f4901ddfc2718656/src/libraries/System.Net.Connections/src/System/Net/Connections/Sockets/SocketsConnectionFactory.cs#L49-L112)). cc: @antonfirsov

We've seen issues in the past where some environments don't properly support IPv6 and/or dual-mode sockets.
@mkellen1 can you please try running your sample with the [`DOTNET_SYSTEM_NET_DISABLEIPV6`](https://devblogs.microsoft.com/dotnet/dotnet-6-networking-improvements/#:~:text=An%20option%20to%20globally%20disable%20IPv6) environment variable set to `1`?

**Comment by mkellen1** [comment]
@MihaZupan Setting that environment variable made no difference in either dotnet-6.0.417 or PowerShell 7.4.0 for my quick test script. If you would like the output from the instrumented test script, I can try and get some time to run it later this week.

**Comment by MihaZupan** [comment]
Traces with the environment variable set on a modern version would be useful to confirm that IPv4 was indeed used

**Comment by mkellen1** [comment]
(Okay, possibly we have more than one issue, as the quick Powershell test fails as before, but the instrumented test program did not. Sticking to the purely dotnet issue for now.)

The test script fails as before in dotnet-8.0.100, but worked fine when I disabled IPv6. 

[YesIPv6-run1.txt](https://github.com/dotnet/runtime/files/13823020/YesIPv6.txt)
[YesIPv6-run2.txt](https://github.com/dotnet/runtime/files/13823076/YesIPv6-2.txt)

[NoIPv6.txt](https://github.com/dotnet/runtime/files/13823022/NoIPv6.txt)

**Comment by wfurt** [comment]
Are you connecting to `localhost` @mkellen1 ????

```
20:09:11.1968790[ConnectionEstablished] versionMajor: 1, versionMinor: 1, connectionId: 0, scheme: http, host: localhost, port: 80, remoteAddress: ::ffff:127.0.0.1
20:09:11.1973118[HandlerMessage] poolId: 30015890, workerId: 11429296, requestId: 0, memberName: TraceConnection, message: HttpConnection(HttpConnectionPool http://localhost:80). Id:0
20:09:11.1973916[HandlerMessage] poolId: 30015890, workerId: 11429296, requestId: 0, memberName: ReturnHttp11Connection, message: isNewConnection=True
20:09:11.1974203[HandlerMessage] poolId: 30015890, workerId: 11429296, requestId: 0, memberName: ReturnHttp11Connection, message: Dequeued waiting HTTP/1.1 request.
20:09:11.1981666[RequestLeftQueue] timeOnQueueMilliseconds: 28.9904, versionMajor: 1, versionMinor: 1
20:09:11.2001363[HandlerMessage] poolId: 30015890, workerId: 11429296, requestId: 41149443, memberName: SendAsync, message: Sending request: Method: GET, RequestUri: 'http://localhost/', Version: 1.1, Content: <null>, Headers:
{
}
20:09:11.2001909[RequestHeadersStart] connectionId: 0
20:09:11.2003014[RequestHeadersStop] 
20:09:11.2003279[HandlerMessage] poolId: 30015890, workerId: 11429296, requestId: 41149443, memberName: WriteToStreamAsync, message: Writing 35 bytes.
20:09:11.2003702[Info] thisOrContextObject: AwaitableSocketAsyncEventArgs#39785641, memberName: InitializeInternals, message: new PreAllocatedOverlapped PreAllocatedOverlapped#45523402
20:09:11.2004954[DumpBuffer] thisOrContextObject: AwaitableSocketAsyncEventArgs#39785641, memberName: LogBuffer, buffer: System.Byte[]
20:09:12.2060514[ErrorMessage] thisOrContextObject: Socket#4032828, memberName: UpdateStatusAfterSocketError, message: errorCode:ConnectionReset, disconnectOnFailure:True
20:09:12.2061192[Info] thisOrContextObject: Socket#4032828, memberName: UpdateStatusAfterSocketError, message: Invalidating socket.
20:09:12.2061373[Info] thisOrContextObject: Socket#4032828, memberName: SetToDisconnected, message: !Disposed
20:09:12.2189224[HandlerMessage] poolId: 30015890, workerId: 11429296, requestId: 41149443, memberName: SendAsync, message: Error sending request: System.IO.IOException: Unable to read data from the transport connection: An existing connection was forcibly closed by the remote host..
 ---> System.Net.Sockets.SocketException (10054): An existing connection was forcibly closed by the remote host.
   --- End of inner exception stack trace ---
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource<System.Int32>.GetResult(Int16 token)
   at System.Net.Http.HttpConnection.InitialFillAsync(Boolean async)
   at System.Net.Http.HttpConnection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
20:09:12.2189779[HandlerMessage] poolId: 30015890, workerId: 11429296, requestId: 41149443, memberName: Dispose, message: Connection closing.
```

This really looks like server is not responding and closing connection. You should be able top see more if you do packet captures.

There is still possibility that something sits on your machine trying to intercept and inspect traffic.

**Comment by mkellen1** [comment]
> @mkellen1 did you mean the powershell test or the .NET test program in [#95487 (comment)](https://github.com/dotnet/runtime/issues/95487#issuecomment-1839304689)?

@antonfirsov That comment referred to testing using the dotnet code in the first post.

**Comment by antonfirsov** [comment]
It is possible that the test exposed the IPv6 issue then, and in https://github.com/dotnet/runtime/issues/95487#issuecomment-1875998814 we may see one of the software mentioned in https://github.com/dotnet/runtime/issues/95487#issuecomment-1839093285 closing the connection. Can you please check whether disabling this SW is changing anything in PowerShell?

**Comment by mkellen1** [comment]
Apologies for the delay, but I finally got some free time to do proper testing.

The issue is definitely related to IPv6. When I run the test dotnet code after setting the environment flag, or Powershell7 code after setting the flag **in my Profile**, connections work just fine. 

Running either with IPv6 on fails, regardless of the state of web filtering software on the device. The first attempt really, truly, honestly does not produce any capturable traffic on ::1 or 127.0.0.1 80/tcp

Here are outputs, Apache logs, and pcaps or screenshots of lack of same:

[Default-8.0.100.zip](https://github.com/dotnet/runtime/files/13924288/Default-8.0.100.zip)
[NoIPv6-8.0.100.zip](https://github.com/dotnet/runtime/files/13924289/NoIPv6-8.0.100.zip)
[NoFilter-8.0.100.zip](https://github.com/dotnet/runtime/files/13924290/NoFilter-8.0.100.zip)

**Comment by mkellen1** [comment]
And yet there was clearly something in the dotnet code changes that is triggering it, as the older versions still work fine. It's not purely environmental, or they would all fail.

**Comment by MihaZupan** [comment]
Yes - https://github.com/dotnet/runtime/issues/95487#issuecomment-1874483944. We started using dual-mode sockets when connecting in 5.0 RC1 and newer.
We currently have no indication that there are issues with our implementation, so as Anton mentioned it appears there's an environmental issue on your side that's not playing nice with IPv6.

If you're unable to track down what's causing issues in your environment, is the workaround sufficient?

**Comment by mkellen1** [comment]
This is a workaround for IPv4 available resources, which is most things. IPv6-only resources are still unreachable.

The older method could handle whatever response it is getting from the system, but the new method chokes on it. That sounds to me like an "indication of an issue".  Obviously, I'm an edge case, so I understand if you don't want to chase it down.

**Comment by antonfirsov** [comment]
What does the following code do with `true`/`false` for `dualMode` with your webserver running on `localhost:80`?

```C#
bool dualMode = true | false; // pls try both
IPAddress serverAddress = dualMode ? IPAddress.Loopback.MapToIPv6() : IPAddress.Loopback;
using Socket socket = dualMode ? new Socket(SocketType.Stream, ProtocolType.Tcp) : new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);

await socket.ConnectAsync(new IPEndPoint(serverAddress, 80));

byte[] requestBytes = Encoding.UTF8.GetBytes("GET / HTTP/1.1\r\nHost: localhost\r\n\r\n");
await socket.SendAsync(requestBytes, SocketFlags.None);

byte[] responseBytes = new byte[1024];
int cnt = await socket.ReceiveAsync(responseBytes, SocketFlags.None);
Console.WriteLine(Encoding.UTF8.GetString(responseBytes, 0, cnt));
```

**Comment by mkellen1** [comment]
When I ran it with `true`, it failed with `An existing connection was forcibly closed by the remote host.` and no packet trace.

When I ran it with `false`, it connected just fine.

**Comment by mkellen1** [comment]
Interestingly, the test case works for a dual-mode connection to `::1` but fails for `:ffff:127.0.0.1`. Is it a requirement to map IPv4 addresses into IPv6?

**Comment by wfurt** [comment]
It should not be @mkellen1. But I think this is sign that something is interfering with the sockets. In the first case, you have IPv6 socket connecting to IPv6 address. In second case are are connecting to IPv4 address.

**Comment by antonfirsov** [comment]
`::1` is an actual IPv6 address, while `:ffff:127.0.0.1` is a [dual-stack](https://learn.microsoft.com/en-us/windows/win32/winsock/dual-stack-sockets) mapping of `127.0.0.1`. This means that your system has no issues with IPv6 in general, but something is shutting down the first connection attempt for dual-stack sockets. This may cause problems in other software using dual-stack. I would try to review the machine configuration once again (are there any special drivers? try to uninstall special networking software, not just disable it, maybe involve your IT department ... etc. )

Unless such issues turn out to affect many users, we have no intention to revert the .NET 5+ change.

**Comment by antonfirsov** [comment]
Triage: keeping this open for a while in case other users would land here and want to comment, moving to Future.

**Comment by mkellen1** [comment]
I wasn't implying that you should roll this back, just wondering why you bother to cast it into a IPv6-mapped form, since that (1) wastes cycles and (2) appears to be what triggers whatever is going on in my environment.

**Comment by wfurt** [comment]
You have to for dual mode sockets as the underlying socket is `AF_INET6`. We used separate sockets in the past but the logic when you connect to name that can resolve to different address families is not trivial. 

We may eventually implement the [happy eyball](https://github.com/dotnet/runtime/issues/87932) and connect to both `AddressFamilies` in parallel. (will use double of system resources and has other problems in certain scenarios)

**Comment by mkellen1** [comment]
Thank you for the explanation! 

I will continue to test with other components of my environment and report back if I ever find a resolution.

**Comment by mkellen1** [comment]
Further testing has revealed that disabling the web filtering client was insufficient. Completely blocking any component of the web filter from running allowed the connections to work correctly without disabling IPv6. I have opened a community thread over there to see if there is a configuration fix.

For the record, the software is Forcepoint's F1E client. (tested on v22)

**Comment by MihaZupan** [comment]
The issue was confirmed to be environmental, closing.
For others running into similar issues, there are useful diagnostic steps described above: https://github.com/dotnet/runtime/issues/95487#issuecomment-1854854890, https://github.com/dotnet/runtime/issues/95487#issuecomment-1874483944, https://github.com/dotnet/runtime/issues/95487#issuecomment-1893929833
